go/client/chat_api_handler.go

```diff
@@ -0,0 +1,199 @@
+}
+
+// Valid returns true if the ChatChannel has at least a Name.
+func (c ChatChannel) Valid() bool {
+	return len(c.Name) > 0
+}
+
+// TopicTypeEnum returns the chat1.TopicType from the TopicType string field.
+func (c ChatChannel) TopicTypeEnum() chat1.TopicType {
+	if len(c.TopicType) == 0 {
+		return chat1.TopicType_CHAT
+	}
+	tt, ok := chat1.TopicTypeMap[strings.ToUpper(c.TopicType)]
+	if ok {
+		return tt
+	}
+
+	return chat1.TopicType_NONE
+}
+
+// ChatMessage represents a text message to be sent.
+type ChatMessage struct {
+	Body string
+}
+
+// Valid returns true if the message has a body.
+func (c ChatMessage) Valid() bool {
+	return len(c.Body) > 0
+}
+
+type sendOptionsV1 struct {
+	Channel        ChatChannel
+	ConversationID chat1.ConversationID `json:"conversation_id"`
+	Message        ChatMessage
+}
+
+func (s sendOptionsV1) Check() error {
+	if !s.Channel.Valid() {
+		return ErrInvalidOptions{version: 1, method: "send", err: errors.New("invalid channel")}
+	}
+	if !s.Message.Valid() {
+		return ErrInvalidOptions{version: 1, method: "send", err: errors.New("invalid message")}
+	}
+
+	return nil
+}
+
+type readOptionsV1 struct {
+	Channel        ChatChannel
+	ConversationID chat1.ConversationID `json:"conversation_id"`
+	Limit          string
+}
+
+func (r readOptionsV1) Check() error {
+	if !r.Channel.Valid() && r.ConversationID == 0 {
+		return ErrInvalidOptions{version: 1, method: "read", err: errors.New("need channel or conversation_id")}
+	}
+	if r.Channel.Valid() && r.ConversationID > 0 {
+		return ErrInvalidOptions{version: 1, method: "read", err: errors.New("include channel or conversation_id, not both")}
+	}
+
+	return nil
+}
+
+func (a *ChatAPI) ListV1(ctx context.Context, c Call, w io.Writer) error {
+	if len(c.Params.Options) != 0 {
+		return ErrInvalidOptions{version: 1, method: "list", err: errors.New("unexpected options, should be empty")}
+	}
+
+	return a.encodeReply(c, a.svcHandler.ListV1(ctx), w)
+}
+
+func (a *ChatAPI) ReadV1(ctx context.Context, c Call, w io.Writer) error {
+	if len(c.Params.Options) == 0 {
+		return ErrInvalidOptions{version: 1, method: "read", err: errors.New("empty options")}
+	}
+	var opts readOptionsV1
+	if err := json.Unmarshal(c.Params.Options, &opts); err != nil {
+		return err
+	}
+	if err := opts.Check(); err != nil {
+		return err
+	}
+
+	// opts are valid for read v1
+
+	return a.encodeReply(c, a.svcHandler.ReadV1(ctx, opts), w)
+}
+
+func (a *ChatAPI) SendV1(ctx context.Context, c Call, w io.Writer) error {
+	if len(c.Params.Options) == 0 {
+		return ErrInvalidOptions{version: 1, method: "send", err: errors.New("empty options")}
+	}
+	var opts sendOptionsV1
+	if err := json.Unmarshal(c.Params.Options, &opts); err != nil {
+		return err
+	}
+	if err := opts.Check(); err != nil {
+		return err
+	}
+
+	// opts are valid for send v1
+
+	return a.encodeReply(c, a.svcHandler.SendV1(ctx, opts), w)
+}
+
+func (a *ChatAPI) encodeReply(call Call, reply Reply, w io.Writer) error {
```

Comment:
Have you considered constructing the `json.Encoder` [here](https://github.com/keybase/client/pull/4252/files#diff-3968c3fab4bd30e2e61ac0c456fd03b4R50), and let this function accept a `*json.Encoder`? This way we can only construct it once and use it for all rpc calls on the same writer. It also make sure the JSON encoder is the only owner of the writer, so that we don't accidentally write other stuff into the writer.